src/terms/local.jl

```diff
@@ -82,78 +82,149 @@ end
-            structure_factor = sum(r -> cis2pi(-dot(G, r)), @view model.positions[group])
-            form_factors[(igroup, q)] * structure_factor
+            structure_factor = sum(r -> cis2pi(-dot(G, r)), @view positions[group])
+            form_factors[(igroup, p)] * structure_factor
         end
         pot / sqrt(model.unit_cell_volume)
     end
 
-    enforce_real!(basis, pot_fourier)  # Symmetrize Fourier coeffs to have real iFFT
-    pot_real = irfft(basis, to_device(basis.architecture, pot_fourier))
-
-    TermAtomicLocal(pot_real)
+    if iszero(q)
+        enforce_real!(basis, pot_fourier)  # Symmetrize coeffs to have real iFFT
+        S = T
+    else
+        S = complex(T)
+    end
+    ifft(S, basis, to_device(basis.architecture, pot_fourier))
 end
+(::AtomicLocal)(basis::PlaneWaveBasis) = TermAtomicLocal(compute_local_potential(basis))
 
-@timing "forces: local" function compute_forces(::TermAtomicLocal, basis::PlaneWaveBasis{TT},
-                                                ψ, occupation; ρ, kwargs...) where {TT}
-    T = promote_type(TT, real(eltype(ψ[1])))
+# Standard computation.
+function compute_forces(term::TermAtomicLocal, basis::PlaneWaveBasis{T}, ψ, occupation;
+                        ρ, kwargs...) where {T}
+    S = promote_type(T, real(eltype(ψ[1])))
+    forces_local(S, basis, ψ, occupation, zero(Vec3{T}), ρ; kwargs...)
+end
+# General function compatible to phonon computations.
+@timing "forces: local" function forces_local(S, basis::PlaneWaveBasis{T}, ψ, occupation, q,
+                                              ρ; kwargs...) where {T}
     model = basis.model
+    recip_lattice = model.recip_lattice
     ρ_fourier = fft(basis, total_density(ρ))
 
     # energy = sum of form_factor(G) * struct_factor(G) * rho(G)
     # where struct_factor(G) = e^{-i G·r}
-    forces = [zero(Vec3{T}) for _ in 1:length(model.positions)]
+    forces = [zero(Vec3{S}) for _ = 1:length(model.positions)]
     for group in model.atom_groups
         element = model.atoms[first(group)]
-        form_factors = [Complex{T}(local_potential_fourier(element, norm(G)))
-                        for G in G_vectors_cart(basis)]
+        form_factors = [complex(S)(local_potential_fourier(element, norm(recip_lattice * (G + q))))
+                        for G in G_vectors(basis)]
         for idx in group
             r = model.positions[idx]
-            forces[idx] = _force_local_internal(basis, ρ_fourier, form_factors, r)
+            forces[idx] = -force_type(S, sum(conj(ρ_fourier[iG])
+                                                 .* form_factors[iG]
+                                                 .* cis2pi(-dot(G + q, r))
+                                                 .* (-2T(π)) .* (G + q) .* im
+                                                 ./ sqrt(basis.model.unit_cell_volume)
+                                             for (iG, G) in enumerate(G_vectors(basis))))
```

Comment:
nice

Reply:
but this is static arrays and scalars, so you might as well not use broadcast and do plain old * and /, shouldn't have a perf impact